### DISCUSSION TITLE
How to utilize Postgres array_append to append to an array with autobase and async session

### DISCUSSION THREAD

**Discussion by JohnnyRacer** [sqlalchemy/sqlalchemy#9592] [Usage Questions] [answered]
I've tried to modify the solution from SO as mentioned in this discussion post [here](https://github.com/sqlalchemy/sqlalchemy/discussions/8799). But it does not appear to work with an autobase model. I am getting this error :
```py
 sqlalchemy.exc.ProgrammingError: (sqlalchemy.dialects.postgresql.asyncpg.ProgrammingError) <class 'asyncpg.exceptions.UndefinedFunctionError'>: function array_append(json[]) does not exist
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
```

The model definition and code snippet I am using is below:
```py
from sqlalchemy.ext.automap import automap_base
from sqlalchemy import ARRAY, JSON, Column, func
Base = automap_base()

class TestUser(Base):
    __tablename__ = 'testdata'
    id = Column('id',Integer,autoincrement=True, primary_key=True)
    time_created = Column('createdAt' ,DateTime(timezone=True), server_default=func.now())
    time_updated = Column('updatedAt' ,DateTime(timezone=True), onupdate=func.now())

    user_id = Column('user_id', String, nullable=False)
    media = Column('data', ARRAY(JSON), nullable=False, default=[])

# async_session is created via sessionmaker(create_async_engine(('CONNECTION_STRING'), echo=True,future=True))

async with async_session() as session:  
    stmt = update(CategoricalContent).where((TestUser.user_id == user_id)).values(tags=func.array_append(TestUser.media))
    res = await session.execute(stmt, {
        image_id : 1234
    })
    await session.commit()
```

Any ideas on how to make this work?

**Comment by CaselIT** [comment]
Hi

Array_append expects two arguments, so you should probably pass the two array to concatenate

**Reply by JohnnyRacer** [reply]
I am not sure how that is possible, I get an error like this if I try your suggestion:
```py
AttributeError: Neither 'Function' object nor 'Comparator' object has an attribute 'items'
```
 Could you provide an snippet of what you are suggesting?

**Comment by zzzeek** [comment, marked as answer]
hi -

automap would not involved here because you have explicit types for the columns in question, assuming "image_id", which is an undefined variable in your example, is actually in your model with a datatype.  Also array_append() takes two arguments, so perhaps that's why you are getting that error message.  the second argument needs to be the JSON value you want to append, and it likely needs to be string-formatted here.

**Reply by JohnnyRacer** [reply]
I get the exact same error with this snippet:
```py
import json

async with async_session() as session:  
    update_dict =  {
        image_id : 1234
    }

    stmt = update(CategoricalContent).where((TestUser.user_id == user_id)).values(tags=func.array_append(TestUser.media, json.dumps(update_dict) ))
    res = await session.execute(stmt,)
    await session.commit()
```

or 

```py
async with async_session() as session:  
    update_dict =  {
        image_id : 1234
    }
    stmt = update(CategoricalContent).where((TestUser.user_id == user_id)).values(tags=func.array_append(TestUser.media))
    res = await session.execute(stmt,json.dumps(update_dict))
    await session.commit()
```
Seems like stringifying the JSON does not make a difference.

**Reply by zzzeek** [reply]
you need to produce string SQL that you can send directly to PostgreSQL which works.  Then I can show you how to render that.

**Reply by JohnnyRacer** [reply]
So instead of using the `update` and `func` function with ORM I need to use a native SQL query? Or do you just need the query to create an example. If you could create a code example that utilizes `func.array_append` (doesn't have to be my specific table example per se)  to append to a list I think that would be enough for me to figure the rest out.

**Reply by JohnnyRacer** [reply]
Also if possible, could the `func` function be made to be easier/clearer to use. I have used `sequelize`  for a Javascript project and simply needed to call `fn('array_append', col('media'), {image_id :1234})` within the update function's dict. It would be very helpful if a function like this could be passed into the `values` attribute of the `update` function to build the statement.

**Reply by zzzeek** [reply]
> So instead of using the `update` and `func` function with ORM I need to use a native SQL query? 

no, PostgreSQL is extremely particular about these queries and I dont know what query will actually work, so we need a model

> Or do you just need the query to create an example. 

example please

> If you could create a code example that utilizes `func.array_append` (doesn't have to be my specific table example per se) to append to a list I think that would be enough for me to figure the rest out.

we typically hope the SQL can be provided since we support the SQLAlchemy library here but not PostgreSQL itself.   however there is no issue I can reproduce, just pass the JSON in as a literal()

```py
import asyncio

from sqlalchemy import ARRAY
from sqlalchemy import Column
from sqlalchemy import engine
from sqlalchemy import func
from sqlalchemy import Integer
from sqlalchemy import JSON
from sqlalchemy import literal
from sqlalchemy import update
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.ext.asyncio import engine
from sqlalchemy.orm import declarative_base

Base = declarative_base()


class TestUser(Base):
    __tablename__ = "testdata"
    id = Column("id", Integer, autoincrement=True, primary_key=True)
    media = Column("data", ARRAY(JSON), nullable=False, default=[])


engine = create_async_engine(
    "postgresql+asyncpg://scott:tiger@localhost/test",
    echo=True,
)


async def go():

    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.drop_all)
        await conn.run_sync(Base.metadata.create_all)

    async with AsyncSession(engine) as session:
        session.add(TestUser(id=1, media=[{"j1": "v1"}, {"j2": "v2"}]))
        await session.commit()

    async with AsyncSession(engine) as session:
        stmt = (
            update(TestUser)
            .where((TestUser.id == 1))
            .values(
                media=func.array_append(
                    TestUser.media, literal({"foo": "bar"}, type_=JSON)
                )
            )
        )
        res = await session.execute(stmt)
        await session.commit()


asyncio.run(go())

```

**Reply by CaselIT** [reply]
tested the above with the other 3 pg dialect and all pass (psycopg, pg8000 and psycopg2)

**Reply by CaselIT** [reply]
pg8000 seems to like it event without literal, the other 3 dialect understandably fail

**Reply by JohnnyRacer** [reply]
Seems like the issue was that I was passing in the `func.array_append` to the `tags` parameter instead of `media` and not using the `literal({...}, type_=JSON)` . I updated my code to reflect those changes and it works. However I can only perform inserts with a single dict. if I modify the code like so : `literal([{"foo": "bar"}, {"foo": "baz"}] , type_=ARRAY(JSON))` with a list of dicts, I get the following error : 
```py
File "/usr/local/lib/python3.8/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 717, in _handle_exception
raise translated_error from error
sqlalchemy.exc.DBAPIError: (sqlalchemy.dialects.postgresql.asyncpg.Error) <class 'asyncpg.exceptions.DataError'>: invalid input for query argument $1: ... (descriptor 'encode' for 'str' objects doesn't apply to a 'list' object)
```
Is there a way I can pass in a list of dicts or do I have to iterate through and update them one by one?

**Reply by CaselIT** [reply]
you can't use `array_append` to append an array. You need to use `array_cat`: https://www.postgresql.org/docs/current/functions-array.html

~~that said I don't think it will fix that error~~ nope that works

**Reply by JohnnyRacer** [reply]
I can confirm it works as well with `
```py 
update(...).where(...).values(media=func.array_cat(TestUser.media, literal([{"foo": "bar"}, {"foo": "baz"}], type_=JSON))
```
The literal is still required for it to work correctly since I am using `psycopg2`. Thanks a lot for all the help :+1:  @CaselIT  @zzzeek

**Reply by CaselIT** [reply]
yes, it's better to use literal for complex types. pg8000 may have some hard coded logic to properly handle these types, but it's probably best not to rely on it

**Reply by zzzeek** [reply]
just noting this is all goofy postgresql stuff.  SQLAlchemy doesn't really "know" about these operations

**Comment by JohnnyRacer** [comment]
I used a dict with the `image_id` key as an example, since in my actual table its receiving JSON with varying and unknown keys. I just want to use `array_append` to append to the `media` list of JSON. A working example snippet would be very helpful as the docs does not mention native array functions with `func`.
